source/robot_model/src/Model.cpp

```diff
@@ -260,9 +262,138 @@ std::vector<state_representation::CartesianPose> Model::forward_geometry(const s
+  Eigen::VectorXd Psi(this->robot_model_.nq);
+  Eigen::VectorXd q = joint_positions.get_positions();
+  for(int i=0; i < this->robot_model_.nq; i++){
+    Psi[i]=0;
+    if(q[i] < this->robot_model_.lowerPositionLimit[i] + margin){
+      Psi[i] = this->robot_model_.upperPositionLimit[i] - margin - std::max(q[i], this->robot_model_.lowerPositionLimit[i]);
+    }
+    else if(this->robot_model_.upperPositionLimit[i] - margin < q[i]){
+      Psi[i] = this->robot_model_.lowerPositionLimit[i] + margin - std::min(q[i], this->robot_model_.upperPositionLimit[i]);
+    }     
+  }
+  
+  return Psi;
+}
+
+void Model::clamp_joint_positions(state_representation::JointPositions& joint_positions) {
+  Eigen::VectorXd q_clamped = joint_positions.get_positions();
+  for(int i=0; i < this->robot_model_.nq; i++){
+    q_clamped[i] = std::min(std::max(q_clamped[i], this->robot_model_.lowerPositionLimit[i]), this->robot_model_.upperPositionLimit[i]);
+  }
+  joint_positions.set_positions(q_clamped);
+}
+
+state_representation::JointPositions Model::inverse_geometry(const state_representation::CartesianState& desired_cartesian_state,
+                                                            const state_representation::JointState& current_joint_state,
+                                                            std::string frame_name,
+                                                            Model::inverse_geometry_parameters params) {
+
+  if (frame_name.empty()) {
+    // get last frame if none specified
+    frame_name = this->robot_model_.frames.back().name;
+  }
+  else {
+    // throw error if specified frame does not exist
+    if (!this->robot_model_.existFrame(frame_name)) {
+      throw (exceptions::FrameNotFoundException(frame_name));
+    }
+  }
+
+  state_representation::JointPositions q(current_joint_state);
+  state_representation::JointVelocities dq(this->get_robot_name(), this->get_number_of_joints());
+
+  const int frame_id = this->robot_model_.getFrameId(frame_name);
+  
+  const pinocchio::SE3 oMdes(desired_cartesian_state.get_orientation().toRotationMatrix(), desired_cartesian_state.get_position());
+  
+  // 1 second pour le Newton-Raphson algorithm
+  std::chrono::nanoseconds dt((int)(1e9));
+
+  pinocchio::Data::Matrix6x J(6,this->robot_model_.nq);
+  pinocchio::Data::Matrix6x J_b(6,this->robot_model_.nq);
+  J.setZero();
+  J_b.setZero();
+  Eigen::VectorXd v(this->robot_model_.nq);
+  Eigen::VectorXd delta_q(this->robot_model_.nq);
+  Eigen::MatrixXd W_b = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  Eigen::MatrixXd W_c = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  Eigen::VectorXd psi(this->robot_model_.nq);
+
+  this->forward_geometry(q, frame_id);
+    
+  // error in SE(3)
+  const pinocchio::SE3 dMf = oMdes.actInv(this->robot_data_.oMf[frame_id]);
+  typedef Eigen::Matrix<double, 6, 1> Vector6d;
+  Vector6d err = pinocchio::log6(dMf).toVector();
```

Comment:
those 3 lines are a repetition of 378 and next. If you need them to initialize your loop then use a `do...while()` syntax

Reply:
Yes. this because I need the error (delta_x) before doing delta_q=J⁻¹delta_x. but I also need it updated as condition for exit from the while loop. The `do while` loop doesn't solve the issue

Reply:
I think the following structure for the loop would be neater:
```c++
while (i < params.max_iterations) {
  // calculate error

  if (err.array().abs().maxCoeff() <= params.tolerance) {
    return q;
  }

  // update q based on algorithm 

  ++i;
}

throw(exceptions::IKDoesNotConverge);
```

Reply:
yes it is better! thanks!